src/MySqlConnector/MySqlClient/ConnectionPool.cs

```diff
@@ -7,101 +9,191 @@ namespace MySql.Data.MySqlClient
 {
 	internal sealed class ConnectionPool
 	{
+
 		public async Task<MySqlSession> TryGetSessionAsync(CancellationToken cancellationToken)
 		{
+			cancellationToken.ThrowIfCancellationRequested();
+
+			// if the drain lock is held, connection draining is in progress
+			await m_drain_lock.WaitAsync(cancellationToken).ConfigureAwait(false);
+			m_drain_lock.Release();
+
+			// wait for an open slot
+			await m_session_semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
 			try
 			{
-				await m_semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+				MySqlSession session;
+				// check for a pooled session
+				if (m_sessions.TryDequeue(out session))
+				{
+					if (!await session.TryPingAsync(cancellationToken).ConfigureAwait(false))
+					{
+						// session is not valid
+						Utility.Dispose(ref session);
+					}
+					else
+					{
+						// session is valid, reset if supported
+						if (m_connectionStringBuilder.ConnectionReset)
+						{
+							await session.ResetConnectionAsync(m_connectionStringBuilder, cancellationToken).ConfigureAwait(false);
+						}
+						// pooled session is ready to be used; return it
+						return session;
+					}
+				}
 
-				if (m_sessions.Count > 0)
-					return m_sessions.Dequeue();
+				session = new MySqlSession(this);
+				await session.ConnectAsync(m_connectionStringBuilder, cancellationToken);
+				return session;
 			}
-			finally
+			catch
```

Comment:
The `m_session_semaphore` is now set to the size of `m_max_connections`.  It is held until the connection is returned to the pool in the Return(session) method.  If there is an exception, it should be returned to the pool immediately since the session never leaves the pool.